lib/winston/config/index.d.ts

```diff
@@ -4,86 +4,26 @@
 /// <reference types="node" />
 
 declare namespace winston {
-  interface AbstractConfigSetLevels {
-    [key: string]: number;
-  }
+  type AbstractConfigSetLevels<L extends string = string> = Record<L, number>;
 
-  interface AbstractConfigSetColors {
-    [key: string]: string | string[];
-  }
+  type AbstractConfigSetColors<L extends string = string> = Record<L, string|string[]>;
 
   interface AbstractConfigSet {
     levels: AbstractConfigSetLevels;
     colors: AbstractConfigSetColors;
   }
 
-  interface CliConfigSetLevels extends AbstractConfigSetLevels {
-    error: number;
-    warn: number;
-    help: number;
-    data: number;
-    info: number;
-    debug: number;
-    prompt: number;
-    verbose: number;
-    input: number;
-    silly: number;
-  }
-
-  interface CliConfigSetColors extends AbstractConfigSetColors {
-    error: string | string[];
-    warn: string | string[];
-    help: string | string[];
-    data: string | string[];
-    info: string | string[];
-    debug: string | string[];
-    prompt: string | string[];
-    verbose: string | string[];
-    input: string | string[];
-    silly: string | string[];
-  }
+  type CliLevels = 'data'|'debug'|'error'|'info'|'input'|'help'|'prompt'|'silly'|'verbose'|'warn';
+  type CliConfigSetLevels = AbstractConfigSetLevels<CliLevels>;
+  type CliConfigSetColors = AbstractConfigSetColors<CliLevels>;
 
-  interface NpmConfigSetLevels extends AbstractConfigSetLevels {
-    error: number;
-    warn: number;
-    info: number;
-    http: number;
-    verbose: number;
-    debug: number;
-    silly: number;
-  }
+  type NpmLevels = 'debug'|'error'|'http'|'info'|'silly'|'verbose'|'warn';
```

Comment:
Wouldn't this:
```suggestion
  type NpmLevels = 'debug'|'error'|'http'|'info'|'silly'|'verbose'|'warn';
```
also be better implemented as:
```typescript
enum NpmLevels {
  Error = 'error',
  Warn = 'warn',
  ...
}
```
The use of a  `type` does limit input values but it also requires that input values be expressed as a string literal everywhere they are used.
```typescript
winston.createLogger({
    level: 'info',
})
```
could be better written as:
```typescript
winston.createLogger({
    level: LogLevel.Info
})
```

Reply:
Using an enum here simply won't do
1. the `enum` would be declared in a declaration file.  
Meaning TypeScript would expect the JS to contain the actual object to use at runtime.  
As the JS doesn't have no enum declared, you're heading for a `TypeError: Cannot read properties of undefined` at runtime
2. `enums` don't provide the flexibility that generics do, which is the whole point of this PR.
3. Based on my own company's codebase, most projects don't even instantiate winston directly.  
We have a wrapper that provides an auto-configured winston instance, custom levels, transports, and all.  
The user may override most every setting, but I find it rare to happen beyond the log level.  
By all means, if you want to use an `enum` in your wrappers to map the levels, be my guest.  

(And more digressions on why using an `enum` here instead of a string union type wouldn't help. Disregarding the above for the sake of the argument)

4. The type as defined is used to generate subsequent types.  
I've grown wary of using `enums` in typings due to the revese linking TypeScript bakes into its `enums` definitions and implementation. (At least, they're consistent!)  
```typescript
// Starting from the simple union type to expose the values
type NpmLevels = 'debug'|'error'|'http'|'info'|'silly'|'verbose'|'warn';

// An enum, based on these values, and your suggestion would have the following typing
const theEnum: {
    [key: Capitalize<NpmLevels>]:  NpmLevels,
    [key: NpmLevels]: Capitalize<NpmLevels>
} = {/* ... */}
// Extracting types from the enum to derive the other types (loglevel config, colors and whatnot)
type keys = keyof typeof theEnum // Capitalize<NpmLevels>|NpmLevels
````

5. Obviously `NpmLevels !== Capitalize<NpmLevels>|NpmLevels`.  
(This still holds true for whatever variation there is between the enum keys and the actual values)  
And there is no way to filter out keys from the first map to the reverse map from the `enum`, as `winston` doesn't filter or transform the logging function names in any way.  
It would be totally valid for someone to use custom levels just to have them capitalized.  
Not sure why anyone would do that, but so far the JS allows for it.
6. The fact that the keyset extended from our intended logging levels, _and their name in the enum_, means the TypeScript typings would fall short as they would allow for invalid values at runtime.  
ie: TypeScript wouldn't raise errors on invoking `logger.Info()` in your code, even if you're using the default npm-based levels.
7. It would be possible to work around these issues by using an enum where the keys are the values (`{[key: T]: T}`), but that would quite defeat the goal of using enums in the fist place.

`enums` are great to map values to keys, but using them as types, as would be the case here, I've found that failscades most every time.